# ***TASK 1: CLASSIFICATION OF MOVIE REVIEWS***

In [1]:
from tensorflow.keras.datasets import imdb
from tensorflow import keras
from tensorflow.keras import layers
(train_data, train_labels), (test_data, test_labels) = imdb.load_data(
    num_words=10000)

17464789/17464789 ━━━━━━━━━━━━━━━━━━━━ 0s 0us/step


In [2]:
import numpy as np
def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        for j in sequence:
            results[i, j] = 1.
    return results
x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [3]:
y_train = np.asarray(train_labels).astype("float32")
y_test = np.asarray(test_labels).astype("float32")

In [4]:
x_val = x_train[:10000]
partial_x_train = x_train[10000:]
y_val = y_train[:10000]
partial_y_train = y_train[10000:]

##Training models with 1, 2, and 3 hidden layers

In [6]:
model = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.fit(partial_x_train,
              partial_y_train,
              epochs=20,
              batch_size=512,
              validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 82ms/step - accuracy: 0.6991 - loss: 0.5792 - val_accuracy: 0.8651 - val_loss: 0.3796
Epoch 2/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 48ms/step - accuracy: 0.9016 - loss: 0.3195 - val_accuracy: 0.8684 - val_loss: 0.3301
Epoch 3/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 33ms/step - accuracy: 0.9249 - loss: 0.2485 - val_accuracy: 0.8844 - val_loss: 0.2955
Epoch 4/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - accuracy: 0.9334 - loss: 0.2066 - val_accuracy: 0.8908 - val_loss: 0.2766
Epoch 5/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9460 - loss: 0.1785 - val_accuracy: 0.8902 - val_loss: 0.2740
Epoch 6/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - accuracy: 0.9528 - loss: 0.1627 - val_accuracy: 0.8814 - val_loss: 0.2969
Epoch 7/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9573 - loss: 0.1402 - val_accuracy: 0.8832 - val_loss: 0.2936
Epoch 8/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 33ms/step - accuracy: 0.9662 - loss: 0.1263 - val_accuracy: 0.8838 - v

In [8]:
model = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.fit(partial_x_train,
              partial_y_train,
              epochs=20,
              batch_size=512,
              validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 61ms/step - accuracy: 0.6869 - loss: 0.6015 - val_accuracy: 0.8499 - val_loss: 0.4090
Epoch 2/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 54ms/step - accuracy: 0.8871 - loss: 0.3429 - val_accuracy: 0.8823 - val_loss: 0.3108
Epoch 3/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 36ms/step - accuracy: 0.9195 - loss: 0.2490 - val_accuracy: 0.8864 - val_loss: 0.2875
Epoch 4/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - accuracy: 0.9333 - loss: 0.2017 - val_accuracy: 0.8806 - val_loss: 0.2958
Epoch 5/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9469 - loss: 0.1622 - val_accuracy: 0.8863 - val_loss: 0.2772
Epoch 6/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 37ms/step - accuracy: 0.9577 - loss: 0.1379 - val_accuracy: 0.8840 - val_loss: 0.2925
Epoch 7/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 36ms/step - accuracy: 0.9627 - loss: 0.1211 - val_accuracy: 0.8790 - val_loss: 0.3030
Epoch 8/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9729 - loss: 0.0999 - val_accuracy: 0.8755 - v

In [10]:
model = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.fit(partial_x_train,
              partial_y_train,
              epochs=20,
              batch_size=512,
              validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 4s 63ms/step - accuracy: 0.6962 - loss: 0.6106 - val_accuracy: 0.8410 - val_loss: 0.4095
Epoch 2/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 37ms/step - accuracy: 0.8892 - loss: 0.3305 - val_accuracy: 0.8854 - val_loss: 0.2959
Epoch 3/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 37ms/step - accuracy: 0.9195 - loss: 0.2314 - val_accuracy: 0.8813 - val_loss: 0.2960
Epoch 4/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - accuracy: 0.9330 - loss: 0.1875 - val_accuracy: 0.8712 - val_loss: 0.3282
Epoch 5/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 32ms/step - accuracy: 0.9547 - loss: 0.1454 - val_accuracy: 0.8828 - val_loss: 0.3056
Epoch 6/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 32ms/step - accuracy: 0.9617 - loss: 0.1212 - val_accuracy: 0.8860 - val_loss: 0.3040
Epoch 7/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9713 - loss: 0.0985 - val_accuracy: 0.8802 - val_loss: 0.3217
Epoch 8/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 47ms/step - accuracy: 0.9775 - loss: 0.0786 - val_accuracy: 0.8814 - v

In [11]:
results

[0.695199191570282, 0.8582000136375427]

###ACCURACY:
####1 Hidden Layer: 0.8658000230789185
####2 Hidden Layers: 0.8595600128173828
####3 Hidden Layers: 0.8582000136375427

##Training models with 32 and 64 layers

In [12]:
model = keras.Sequential([
    layers.Dense(32, activation="relu"),
    layers.Dense(32, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.fit(partial_x_train,
              partial_y_train,
              epochs=20,
              batch_size=512,
              validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 8s 215ms/step - accuracy: 0.6780 - loss: 0.5838 - val_accuracy: 0.8164 - val_loss: 0.4201
Epoch 2/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 6s 82ms/step - accuracy: 0.8903 - loss: 0.3064 - val_accuracy: 0.8786 - val_loss: 0.3051
Epoch 3/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 54ms/step - accuracy: 0.9165 - loss: 0.2308 - val_accuracy: 0.8858 - val_loss: 0.2893
Epoch 4/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 43ms/step - accuracy: 0.9397 - loss: 0.1740 - val_accuracy: 0.8835 - val_loss: 0.2988
Epoch 5/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 52ms/step - accuracy: 0.9511 - loss: 0.1434 - val_accuracy: 0.8861 - val_loss: 0.2876
Epoch 6/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 41ms/step - accuracy: 0.9601 - loss: 0.1230 - val_accuracy: 0.8853 - val_loss: 0.3005
Epoch 7/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 51ms/step - accuracy: 0.9766 - loss: 0.0873 - val_accuracy: 0.8848 - val_loss: 0.3232
Epoch 8/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 60ms/step - accuracy: 0.9780 - loss: 0.0791 - val_accuracy: 0.8843 - 

In [14]:
model = keras.Sequential([
    layers.Dense(64, activation="relu"),
    layers.Dense(64, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.fit(partial_x_train,
              partial_y_train,
              epochs=20,
              batch_size=512,
              validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 5s 127ms/step - accuracy: 0.6721 - loss: 0.5946 - val_accuracy: 0.8712 - val_loss: 0.3551
Epoch 2/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 63ms/step - accuracy: 0.8941 - loss: 0.3013 - val_accuracy: 0.8723 - val_loss: 0.3108
Epoch 3/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 66ms/step - accuracy: 0.9108 - loss: 0.2319 - val_accuracy: 0.8783 - val_loss: 0.2965
Epoch 4/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 67ms/step - accuracy: 0.9379 - loss: 0.1743 - val_accuracy: 0.8826 - val_loss: 0.2954
Epoch 5/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 62ms/step - accuracy: 0.9500 - loss: 0.1436 - val_accuracy: 0.8792 - val_loss: 0.3077
Epoch 6/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 4s 100ms/step - accuracy: 0.9650 - loss: 0.1110 - val_accuracy: 0.8472 - val_loss: 0.4384
Epoch 7/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 62ms/step - accuracy: 0.9669 - loss: 0.0990 - val_accuracy: 0.8813 - val_loss: 0.3276
Epoch 8/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 60ms/step - accuracy: 0.9759 - loss: 0.0782 - val_accuracy: 0.8775 -

In [15]:
results

[0.6514740586280823, 0.8646000027656555]

###ACCURACY:
####16 Units: 0.8595600128173828
####32 Units: 0.8610000014305115
####64 Units: 0.8646000027656555

##Training models with different loss functions

In [16]:
model = keras.Sequential([
    layers.Dense(16, activation="relu"),
    layers.Dense(16, activation="relu"),
    layers.Dense(1, activation="sigmoid")
])
model.compile(optimizer="rmsprop",
              loss="mse",
              metrics=["accuracy"])
model.fit(partial_x_train,
              partial_y_train,
              epochs=20,
              batch_size=512,
              validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 64ms/step - accuracy: 0.6755 - loss: 0.2186 - val_accuracy: 0.8492 - val_loss: 0.1397
Epoch 2/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 38ms/step - accuracy: 0.8781 - loss: 0.1207 - val_accuracy: 0.8697 - val_loss: 0.1081
Epoch 3/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 57ms/step - accuracy: 0.9103 - loss: 0.0836 - val_accuracy: 0.8566 - val_loss: 0.1065
Epoch 4/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 44ms/step - accuracy: 0.9213 - loss: 0.0717 - val_accuracy: 0.8861 - val_loss: 0.0865
Epoch 5/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 35ms/step - accuracy: 0.9353 - loss: 0.0584 - val_accuracy: 0.8863 - val_loss: 0.0837
Epoch 6/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 36ms/step - accuracy: 0.9514 - loss: 0.0471 - val_accuracy: 0.8837 - val_loss: 0.0844
Epoch 7/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 33ms/step - accuracy: 0.9543 - loss: 0.0421 - val_accuracy: 0.8841 - val_loss: 0.0836
Epoch 8/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9644 - loss: 0.0363 - val_accuracy: 0.8737 - v

In [17]:
results

[0.1066281870007515, 0.8651999831199646]

###ACCURACY:
####binary_crossentropy: 0.8595600128173828
####mse: 0.8651999831199646

##Training models with different activation functions

In [18]:
model = keras.Sequential([
    layers.Dense(16, activation="tanh"),
    layers.Dense(16, activation="tanh"),
    layers.Dense(1, activation="sigmoid")
])
model.compile(optimizer="rmsprop",
              loss="binary_crossentropy",
              metrics=["accuracy"])
model.fit(partial_x_train,
              partial_y_train,
              epochs=20,
              batch_size=512,
              validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 3s 64ms/step - accuracy: 0.7147 - loss: 0.5875 - val_accuracy: 0.8586 - val_loss: 0.3956
Epoch 2/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 36ms/step - accuracy: 0.8890 - loss: 0.3397 - val_accuracy: 0.8794 - val_loss: 0.3084
Epoch 3/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9173 - loss: 0.2359 - val_accuracy: 0.8892 - val_loss: 0.2763
Epoch 4/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 35ms/step - accuracy: 0.9394 - loss: 0.1801 - val_accuracy: 0.8902 - val_loss: 0.2719
Epoch 5/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 34ms/step - accuracy: 0.9592 - loss: 0.1352 - val_accuracy: 0.8862 - val_loss: 0.2901
Epoch 6/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 2s 46ms/step - accuracy: 0.9642 - loss: 0.1123 - val_accuracy: 0.8868 - val_loss: 0.3073
Epoch 7/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 48ms/step - accuracy: 0.9714 - loss: 0.0907 - val_accuracy: 0.8814 - val_loss: 0.3447
Epoch 8/20
30/30 ━━━━━━━━━━━━━━━━━━━━ 1s 39ms/step - accuracy: 0.9795 - loss: 0.0732 - val_accuracy: 0.8711 - v

In [19]:
results

[0.7420349717140198, 0.8560799956321716]

###ACCURACY:
####relu: 0.8595600128173828
####tanh: 0.8560799956321716

#***TASK 2: CLASSIFYING NEWSWIRES***

In [69]:
from tensorflow.keras.datasets import reuters
from tensorflow import keras
from tensorflow.keras import layers

##Training with 10%, 50%, and 100% of data

In [70]:
(train_data, train_labels), (test_data, test_labels) = reuters.load_data(num_words=10000)

In [71]:
train_data = train_data[:int(len(train_data)*.1)]
train_labels = train_labels[:int(len(train_labels)*.1)]
test_data = test_data[:int(len(test_data)*.1)]
test_labels = test_labels[:int(len(test_labels)*.1)]

In [46]:
train_data = train_data[:int(len(train_data)*.5)]
train_labels = train_labels[:int(len(train_labels)*.5)]
test_data = test_data[:int(len(test_data)*.5)]
test_labels = test_labels[:int(len(test_labels)*.5)]

In [4]:
train_labels[10]

3

In [72]:
import numpy as np
def vectorize_sequences(sequences, dimension=10000):
    results = np.zeros((len(sequences), dimension))
    for i, sequence in enumerate(sequences):
        for j in sequence:
            results[i, j] = 1.
    return results
x_train = vectorize_sequences(train_data)
x_test = vectorize_sequences(test_data)

In [73]:
def to_one_hot(labels, dimension=46):
    results = np.zeros((len(labels), dimension))
    for i, label in enumerate(labels):
        results[i, label] = 1.
    return results
y_train = to_one_hot(train_labels)
y_test = to_one_hot(test_labels)

VALIDATION SETS

In [74]:
x_val = x_train[:100]
partial_x_train = x_train[100:]
y_val = y_train[:100]
partial_y_train = y_train[100:]

In [49]:
x_val = x_train[:500]
partial_x_train = x_train[500:]
y_val = y_train[:500]
partial_y_train = y_train[500:]

In [56]:
x_val = x_train[:1000]
partial_x_train = x_train[1000:]
y_val = y_train[:1000]
partial_y_train = y_train[1000:]

TRAINING THE MODEL

In [75]:
from tensorflow.keras.utils import to_categorical
y_train = to_categorical(train_labels, num_classes = 46)
y_test = to_categorical(test_labels, num_classes = 46)

In [16]:
len(x_test)

1123

In [17]:
len(y_test)

1123

In [78]:
model = keras.Sequential([
  layers.Dense(64, activation="relu"),
  layers.Dense(64, activation="relu"),
  layers.Dense(46, activation="softmax")
])
model.compile(optimizer="rmsprop",
              loss="categorical_crossentropy",
              metrics=["accuracy"])
model.fit(partial_x_train,
                    partial_y_train,
                    epochs=20,
                    batch_size=512,
                    validation_data=(x_val, y_val))
results = model.evaluate(x_test, y_test)

Epoch 1/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 1s 272ms/step - accuracy: 0.0763 - loss: 3.8042 - val_accuracy: 0.3700 - val_loss: 3.5207
Epoch 2/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step - accuracy: 0.3463 - loss: 3.5185 - val_accuracy: 0.5800 - val_loss: 3.1226
Epoch 3/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 89ms/step - accuracy: 0.4760 - loss: 3.1074 - val_accuracy: 0.5700 - val_loss: 2.7455
Epoch 4/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step - accuracy: 0.5438 - loss: 2.7110 - val_accuracy: 0.6000 - val_loss: 2.3741
Epoch 5/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 84ms/step - accuracy: 0.5988 - loss: 2.3086 - val_accuracy: 0.6100 - val_loss: 2.0510
Epoch 6/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 93ms/step - accuracy: 0.6319 - loss: 1.9695 - val_accuracy: 0.6100 - val_loss: 1.8177
Epoch 7/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 101ms/step - accuracy: 0.6706 - loss: 1.6786 - val_accuracy: 0.6300 - val_loss: 1.6561
Epoch 8/20
2/2 ━━━━━━━━━━━━━━━━━━━━ 0s 86ms/step - accuracy: 0.6979 - loss: 1.4919 - val_accuracy: 0.6800 - val_loss: 1.526

In [79]:
results

[1.4883471727371216, 0.6964285969734192]

###Accuracy (20 Epochs):
####10% Size: 0.6964285969734192
####50% Size: 0.7515583038330078
####100% Size: 0.790739119052887

#***TASK 3: PREDICTING THE HOUSE PRICES***

In [41]:
from tensorflow.keras.datasets import boston_housing
(train_data, train_targets), (test_data, test_targets) = boston_housing.load_data()

In [42]:
mean = train_data.mean(axis=0)
train_data -= mean
std = train_data.std(axis=0)
train_data /= std
test_data -= mean
test_data /= std

In [43]:
def build_model():
    model = keras.Sequential([
        layers.Dense(64, activation="relu"),
        layers.Dense(64, activation="relu"),
        layers.Dense(1)
    ])
    model.compile(optimizer="rmsprop", loss="mse", metrics=["mae"])
    return model

In [44]:
k = 16
num_val_samples = len(train_data) // k
num_epochs = 100
all_scores = []
for i in range(k):
    print(f"Processing fold #{i}")
    val_data = train_data[i * num_val_samples: (i + 1) * num_val_samples]
    val_targets = train_targets[i * num_val_samples: (i + 1) * num_val_samples]
    partial_train_data = np.concatenate(
        [train_data[:i * num_val_samples],
         train_data[(i + 1) * num_val_samples:]],
        axis=0)
    partial_train_targets = np.concatenate(
        [train_targets[:i * num_val_samples],
         train_targets[(i + 1) * num_val_samples:]],
        axis=0)
    model = build_model()
    model.fit(partial_train_data, partial_train_targets,
              epochs=num_epochs, batch_size=16, verbose=0)
    val_mse, val_mae = model.evaluate(val_data, val_targets, verbose=0)
    all_scores.append(val_mae)
test_mse_score, test_mae_score = model.evaluate(test_data, test_targets)

Processing fold #0
Processing fold #1
Processing fold #2
Processing fold #3
Processing fold #4
Processing fold #5
Processing fold #6
Processing fold #7
Processing fold #8
Processing fold #9
Processing fold #10
Processing fold #11
Processing fold #12
Processing fold #13
Processing fold #14
Processing fold #15
4/4 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 16.6263 - mae: 2.8482 


In [45]:
test_mae_score

3.004159927368164

###Acurracy (Mean Absolute Error):
####k = 2: 3.076073169708252
####k = 4: 2.8106303215026855
####k = 8: 2.6309590339660645
####k = 16: 3.004159927368164